src/internal/pfsdb/commits.go

```diff
@@ -542,6 +548,31 @@ func ForEachCommitAncestor(ctx context.Context, extCtx sqlx.ExtContext, startId
 	return nil
 }
 
+// forEachCommitAncestorUntilRoot calls ForEachCommitAncestor continuously until the root is encountered.
+func forEachCommitAncestorUntilRoot(ctx context.Context, tx *pachsql.Tx, startId CommitID, cb func(parentId, childId CommitID) error) error {
+	commitPtr := startId
+	earliest := commitPtr
+	stop := false
+	for {
+		if err := ForEachCommitAncestor(ctx, tx, commitPtr, MaxSearchDepth, func(parentId, childId CommitID) error {
+			earliest = parentId
+			if err := cb(parentId, childId); err != nil {
+				if err == errutil.ErrBreak {
+					stop = true
+				}
+				return err
+			}
+			return nil
+		}); err != nil {
+			return errors.Wrap(err, "for each commit ancestor in batches")
+		}
+		if earliest == commitPtr || stop { // root was found or errutil.ErrBreak was encountered.
```

Comment:
You might be able to get rid of stop here by adding `errutil.ErrBreak` support to `ForEachCommitAncestor`.  This termination seems a little clunky, it took a minute to reason through stop.